### PR TITLE
carto: use buildNpmPackage

### PR DESCRIPTION
**Pull request by dotlambda**
## Description of changes
See https://github.com/NixOS/nixpkgs/issues/229475

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

cc @Luflosi who originally packaged it in https://github.com/NixOS/nixpkgs/pull/187891

### CODE REVIEW

**Review by delroth — APPROVED** [review]
Tested:
- Builds fine on x86_64-linux
- `--version` works
- Diff with previous version:
  - ~~The output from this PR has a `bin/carto`, which is a good improvement.~~ EDIT: that's my `find` invocation not following symlinks, ignore :)
  - However the man page isn't properly copied to `share/man/carto.1` anymore. That's likely a `buildNpmPackage` bug/missing feature? I don't really see that as a major issue though.

File: pkgs/build-support/node/build-npm-package/hooks/npm-install-hook.sh [hunk 1]
```diff
@@ -25,8 +25,7 @@ npmInstallHook() {
         else "invalid type " + $typ | halt_error end' "${npmWorkspace-.}/package.json")
 
     while IFS= read -r man; do
-        mkdir -p "$out/share/man"
-        ln -s "$packageOut/$man" "$out/share/man"
+        installManPage "$packageOut/$man"
```
Comment (dotlambda) [inline pkgs/build-support/node/build-npm-package/hooks/npm-install-hook.sh:28]: This isn't quite accurate either. According to https://docs.npmjs.com/cli/v9/configuring-npm/package-json/#man we're supposed to modify the name of the man page. That would be easiest to do if we modify `installManPage` to accept the man page name as an optional argument.

Comment (delroth) [reply pkgs/build-support/node/build-npm-package/hooks/npm-install-hook.sh:28]: TIL about "installManPage", thanks! No objection to leaving that bug for now, it's still better than the previous or current status and I don't want to block this particular PR on the man page question.

**Review by delroth — APPROVED** [review]
Still LGTM. Leaving another day or so for the maintainer to possibly comment on this, but I think it's ready to merge.

File: pkgs/development/tools/carto/default.nix [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    hash = "sha256-TylMgb2EI52uFmVeMJiQltgNCSh6MutFwUjsYC7gfEA=";
+  };
+
+  npmDepsHash = "sha256-8M9hze71bQWhyxcXeI/EOr0SQ+tx8Lb9LfvnGxYYo0A=";
+
+  dontNpmBuild = true;
+
+  meta = {
+    changelog = "https://github.com/mapbox/carto/blob/${src.rev}/CHANGELOG.md";
+    description = "Fast CSS-like map stylesheets";
```
Comment (Luflosi) [inline pkgs/development/tools/carto/default.nix:23]: While this is in the GitHub "About" section, it is terrible at describing what this program actually does. I propose we keep the old description:
```suggestion
    description = "Mapnik Stylesheet Compiler";
```

File: pkgs/development/tools/carto/default.nix [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    homepage = "https://github.com/mapbox/carto";
+    license = lib.licenses.asl20;
+    mainProgram = "carto";
```
Comment (Luflosi) [inline pkgs/development/tools/carto/default.nix:26]: Both the package and the binary it provides are called "carto", right? In this case you should remove `mainProgram` as it is unnecessary.
```suggestion
```

File: pkgs/build-support/node/build-npm-package/hooks/default.nix [hunk 4]
```diff
@@ -27,7 +27,10 @@
   npmInstallHook = makeSetupHook
     {
       name = "npm-install-hook";
-      propagatedBuildInputs = [ buildPackages.makeWrapper ];
+      propagatedBuildInputs = [
+        buildPackages.installShellFiles
+        buildPackages.makeWrapper
+      ];
```
Comment (Luflosi) [inline pkgs/build-support/node/build-npm-package/hooks/default.nix:33]: I think this looks nicer:
```suggestion
      propagatedBuildInputs = with buildPackages; [
        installShellFiles
        makeWrapper
      ];
```

**Review by Luflosi — COMMENTED** [review]
Besides a few nitpicks, LGTM.

File: pkgs/development/tools/carto/default.nix (hunk 3 above)
Comment (dotlambda) [reply pkgs/development/tools/carto/default.nix:26]: Not anymore: https://github.com/NixOS/nixpkgs/pull/246386

Comment (Luflosi) [reply pkgs/development/tools/carto/default.nix:26]: Thanks, I didn't know.

**Review by Luflosi — APPROVED** [review]
LGTM.

**Comment by dotlambda** [comment]
Result of `nixpkgs-review pr 249879` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>magnetophonDSP.ConstantDetuneChorus</li>
  </ul>
</details>
<details>
  <summary>6 packages failed to build:</summary>
  <ul>
    <li>geph.gui</li>
    <li>imgbrd-grabber</li>
    <li>iosevka-comfy.comfy</li>
    <li>iosevka-comfy.comfy-duo</li>
    <li>iosevka-comfy.comfy-motion</li>
    <li>sitespeed-io</li>
  </ul>
</details>
<details>
  <summary>150 packages built:</summary>
  <ul>
    <li>activitywatch</li>
    <li>ansible-language-server</li>
    <li>antennas</li>
    <li>antora</li>
    <li>ariang</li>
    <li>assemblyscript</li>
    <li>audiobookshelf</li>
    <li>authelia</li>
    <li>aw-server-rust</li>
    <li>balanceofsatoshis</li>
    <li>bibtex-tidy</li>
    <li>bitwarden</li>
    <li>bitwarden-cli</li>
    <li>carto</li>
    <li>castnow</li>
    <li>cdxgen</li>
    <li>cz-cli</li>
    <li>deltachat-desktop</li>
    <li>djot-js</li>
    <li>docker-compose-language-service</li>
    <li>documenso</li>
    <li>dot-language-server</li>
    <li>eask</li>
    <li>elmPackages.elm-test</li>
    <li>emscripten</li>
    <li>epgstation</li>
    <li>eslint_d</li>
    <li>faust (faust2)</li>
    <li>faust2alqt</li>
    <li>faust2alsa</li>
    <li>faust2csound</li>
    <li>faust2firefox</li>
    <li>faust2jack</li>
    <li>faust2jackrust</li>
    <li>faust2jaqt</li>
    <li>faust2ladspa</li>
    <li>faust2lv2</li>
    <li>faust2sc</li>
    <li>faustPhysicalModeling</li>
    <li>faustlive</li>
    <li>filebrowser</li>
    <li>flood</li>
    <li>forgejo</li>
    <li>forgejo.data</li>
    <li>fx-cast-bridge</li>
    <li>github-copilot-cli</li>
    <li>gnomeExtensions.pop-shell</li>
    <li>google-clasp</li>
    <li>gtop</li>
    <li>guitarix</li>
    <li>haskellPackages.aeson-typescript</li>
    <li>haskellPackages.aeson-typescript.doc</li>
    <li>homepage-dashboard</li>
    <li>hred</li>
    <li>hueadm</li>
    <li>husky</li>
    <li>iosevka</li>
    <li>iosevka-comfy.comfy-fixed</li>
    <li>iosevka-comfy.comfy-motion-duo</li>
    <li>iosevka-comfy.comfy-motion-fixed</li>
    <li>iosevka-comfy.comfy-wide</li>
    <li>iosevka-comfy.comfy-wide-duo</li>
    <li>iosevka-comfy.comfy-wide-fixed</li>
    <li>iosevka-comfy.comfy-wide-motion</li>
    <li>iosevka-comfy.comfy-wide-motion-duo</li>
    <li>iosevka-comfy.comfy-wide-motion-fixed</li>
    <li>jellyfin</li>
    <li>jellyfin-media-player</li>
    <li>jellyfin-web</li>
    <li>kapitonov-plugins-pack</li>
    <li>karma-runner</li>
    <li>kaufkauflist</li>
    <li>kavita</li>
    <li>lineselect</li>
    <li>lv_img_conv</li>
    <li>magnetophonDSP.CharacterCompressor</li>
    <li>magnetophonDSP.CompBus</li>
    <li>magnetophonDSP.LazyLimiter</li>
    <li>magnetophonDSP.MBdistortion</li>
    <li>magnetophonDSP.RhythmDelay</li>
    <li>magnetophonDSP.VoiceOfFaust</li>
    <li>magnetophonDSP.faustCompressors</li>
    <li>magnetophonDSP.pluginUtils</li>
    <li>magnetophonDSP.shelfMultiBand</li>
    <li>maizzle</li>
    <li>markdownlint-cli</li>
    <li>matrix-alertmanager</li>
    <li>memos</li>
    <li>mermaid-filter</li>
    <li>mirakurun</li>
    <li>mongosh</li>
    <li>mooSpace</li>
    <li>moonfire-nvr</li>
    <li>mpvScripts.webtorrent-mpv-hook</li>
    <li>mw</li>
    <li>nest-cli</li>
    <li>newman</li>
    <li>node-manta</li>
    <li>nodehun</li>
    <li>npm-check</li>
    <li>ntfy-sh</li>
    <li>open-music-kontrollers.mephisto</li>
    <li>open-stage-control</li>
    <li>osmtogeojson</li>
    <li>pairdrop</li>
    <li>paperless-ngx</li>
    <li>photofield</li>
    <li>photoprism</li>
    <li>pnpm-lock-export</li>
    <li>polaris</li>
    <li>polaris-web</li>
    <li>psitransfer</li>
    <li>pufferpanel</li>
    <li>pyCA</li>
    <li>pyCA.dist</li>
    <li>readability-cli</li>
    <li>resumed</li>
    <li>reveal-md</li>
    <li>semantic-release</li>
    <li>sharing</li>
    <li>slskd</li>
    <li>sunshine</li>
    <li>swagger-cli</li>
    <li>tambura</li>
    <li>terminal-stocks</li>
    <li>thelounge</li>
    <li>torq</li>
    <li>triton</li>
    <li>twspace-crawler</li>
    <li>typescript</li>
    <li>uivonim</li>
    <li>uptime-kuma</li>
    <li>vencord</li>
    <li>vencord-web-extension</li>
    <li>vieb</li>
    <li>vimPlugins.YouCompleteMe</li>
    <li>vsce</li>
    <li>vscode-extensions.vadimcn.vscode-lldb</li>
    <li>vscode-langservers-extracted</li>
    <li>webcord</li>
    <li>webcord-vencord</li>
    <li>webtorrent_desktop</li>
    <li>weylus</li>
    <li>windmill</li>
    <li>yarn-lock-converter</li>
    <li>ycmd</li>
    <li>zigbee2mqtt</li>
    <li>zincsearch</li>
    <li>zrythm</li>
    <li>zx</li>
  </ul>
</details>

**Pull request merged**